File: vennon/self/skills/webview/mermaid/mermaid-live-server.mjs
#!/usr/bin/env node
/**
 * Servidor mínimo para pré-visualização Mermaid em tempo quase real.
 *
 * - GET  /mermaid-live  → SSE: envia o ficheiro observado ao ligar e sempre que mudar
 * - POST /mermaid-push → corpo text/plain = código Mermaid; grava no ficheiro e notifica clientes
 * - GET  /*             → ficheiros estáticos a partir de --static
 *
 * Uso:
 *   node mermaid-live-server.mjs --file ./diagram.mmd --static . --port 9876
 *
 * No browser (mesmo origin que este servidor):
 *   abrir base.html → liga a /mermaid-live por defeito (?nolive=1 desliga); redesenha ao receber texto
 *
 * O agente (ou tu) pode atualizar o gráfico sem refresh manual:
 *   curl -sS -X POST http://127.0.0.1:9876/mermaid-push --data-binary @diagram.mmd
 *   # ou editar diagram.mmd no disco → fs.watch notifica
 */
import http from 'http';
import fs from 'fs';
import path from 'path';

let watchFile = path.join(process.cwd(), 'diagram.mmd');
let port = 9876;
let staticRoot = process.cwd();

for (let i = 2; i < process.argv.length; i++) {
  const a = process.argv[i];
  if (a === '--file' && process.argv[i + 1]) {
    watchFile = path.resolve(process.argv[++i]);
    continue;
  }
  if (a === '--port' && process.argv[i + 1]) {
    port = parseInt(process.argv[++i], 10);
    continue;
  }
  if (a === '--static' && process.argv[i + 1]) {
    staticRoot = path.resolve(process.argv[++i]);
    continue;
  }
  if (a === '-h' || a === '--help') {
    console.log(`Usage: node mermaid-live-server.mjs [--file PATH] [--port N] [--static DIR]`);
    process.exit(0);
  }
}

const clients = new Set();
let watchTimer = null;

function broadcast(text) {
  const payload = JSON.stringify({ text: text || '' });
  const chunk = 'data: ' + payload.replace(/\r\n/g, '\n').split('\n').join('\ndata: ') + '\n\n';
  for (const res of clients) {
    try {
      res.write(chunk);
    } catch {
      clients.delete(res);
    }
  }
}

function readAndBroadcast() {
  fs.readFile(watchFile, 'utf8', (err, text) => {
    if (err && err.code !== 'ENOENT') console.error('read', watchFile, err.message);
    broadcast(err ? '' : text);
  });
}

function scheduleBroadcast() {
  if (watchTimer) clearTimeout(watchTimer);
  watchTimer = setTimeout(() => {
    watchTimer = null;
    readAndBroadcast();
  }, 120);
}

const MIME = {
  '.html': 'text/html; charset=utf-8',
  '.js': 'text/javascript; charset=utf-8',
  '.css': 'text/css; charset=utf-8',
  '.mmd': 'text/plain; charset=utf-8',
  '.svg': 'image/svg+xml',
  '.json': 'application/json; charset=utf-8',
};

function safeJoin(root, reqPath) {
  const rel = String(reqPath || '').replace(/^\/+/, '');
  if (rel.includes('\0')) return null;
  const resolved = path.resolve(root, rel);
  if (!resolved.startsWith(path.resolve(root))) return null;
  return resolved;
}

const server = http.createServer((req, res) => {
  const u = new URL(req.url || '/', `http://127.0.0.1:${port}`);

  if (u.pathname === '/mermaid-live' && req.method === 'GET') {
    res.writeHead(200, {
      'Content-Type': 'text/event-stream; charset=utf-8',
      'Cache-Control': 'no-cache, no-transform',
      Connection: 'keep-alive',
      'Access-Control-Allow-Origin': '*',
    });
    clients.add(res);
    readAndBroadcast();
    req.on('close', () => {
      clients.delete(res);
    });
    return;
  }

  if (u.pathname === '/mermaid-push' && req.method === 'POST') {
    let body = '';
    req.on('data', (c) => {
      body += c;
      if (body.length > 2_000_000) req.destroy();
    });
    req.on('end', () => {
      fs.mkdir(path.dirname(watchFile), { recursive: true }, (mkErr) => {
        if (mkErr) {
          res.writeHead(500);
          res.end(mkErr.message);
          return;
        }
        fs.writeFile(watchFile, body, 'utf8', (err) => {
          if (err) {
            res.writeHead(500);
            res.end(err.message);
            return;
          }
          broadcast(body);
          res.writeHead(204, { 'Access-Control-Allow-Origin': '*' });
          res.end();
        });
      });
    });
    return;
  }

  if (req.method !== 'GET' && req.method !== 'HEAD') {
    res.writeHead(405);
    res.end();
    return;
  }

  let rel = u.pathname === '/' ? '/index.html' : u.pathname;
  const filePath = safeJoin(staticRoot, decodeURIComponent(rel));
  if (!filePath) {
    res.writeHead(403);
    res.end('Forbidden');
    return;
  }

  fs.stat(filePath, (err, st) => {
    if (err || !st.isFile()) {
      res.writeHead(404);
      res.end('Not found');
      return;
    }
    const ext = path.extname(filePath).toLowerCase();
    res.writeHead(200, { 'Content-Type': MIME[ext] || 'application/octet-stream' });
    if (req.method === 'HEAD') {
      res.end();
      return;
    }
    fs.createReadStream(filePath).pipe(res);
  });
});

function ensureWatch() {
  fs.mkdir(path.dirname(watchFile), { recursive: true }, () => {
    fs.access(watchFile, fs.constants.F_OK, (err) => {
      if (err) {
        fs.writeFile(
          watchFile,
          "%%{init: {'theme': 'dark'}}%%\nflowchart TD\n  A[Live] --> B[OK]\n",
          'utf8',
          () => tryWatch()
        );
      } else {
        tryWatch();
      }
    });
  });
}

function tryWatch() {
  try {
    fs.watch(watchFile, { persistent: true }, () => scheduleBroadcast());
  } catch (e) {
    console.error('watch', watchFile, e.message);
  }
}

ensureWatch();

server.listen(port, '127.0.0.1', () => {
  console.log(`mermaid-live http://127.0.0.1:${port}/  file=${watchFile}  static=${staticRoot}`);
  console.log(`  SSE GET /mermaid-live   POST /mermaid-push   abrir base.html (?nolive=1 desliga SSE)`);
});

File: vennon/self/skills/webview/mermaid/mermaid_live_server.py
#!/usr/bin/env python3
"""Espelho em Python de mermaid-live-server.mjs (sem dependência Node)."""
from __future__ import annotations

import argparse
import json
import os
import threading
import time
from http.server import ThreadingHTTPServer, SimpleHTTPRequestHandler
from pathlib import Path
from typing import ClassVar
from urllib.parse import urlparse

MIME = {
    ".html": "text/html; charset=utf-8",
    ".js": "text/javascript; charset=utf-8",
    ".css": "text/css; charset=utf-8",
    ".mmd": "text/plain; charset=utf-8",
    ".svg": "image/svg+xml",
    ".json": "application/json; charset=utf-8",
}


class LiveHandler(SimpleHTTPRequestHandler):
    server_version = "MermaidLive/1.0"
    clients: ClassVar[set] = set()
    clients_lock: ClassVar[threading.Lock] = threading.Lock()
    watch_file: ClassVar[Path] = Path("diagram.mmd")
    last_mtime: ClassVar[float | None] = None
    last_sent: ClassVar[str] = ""

    def log_message(self, fmt: str, *args) -> None:
        return

    def _broadcast(self, text: str) -> None:
        payload = json.dumps({"text": text or ""})
        chunk = f"data: {payload}\n\n".encode()
        with LiveHandler.clients_lock:
            dead = []
            for wfile in LiveHandler.clients:
                try:
                    wfile.write(chunk)
                    wfile.flush()
                except OSError:
                    dead.append(wfile)
            for w in dead:
                LiveHandler.clients.discard(w)

    def _read_file(self) -> str:
        p = LiveHandler.watch_file
        try:
            return p.read_text(encoding="utf-8")
        except OSError:
            return ""

    def do_POST(self) -> None:
        u = urlparse(self.path)
        if u.path != "/mermaid-push":
            self.send_error(404)
            return
        n = int(self.headers.get("Content-Length", "0"))
        body = self.rfile.read(min(n, 2_000_000)).decode("utf-8", errors="replace")
        p = LiveHandler.watch_file
        p.parent.mkdir(parents=True, exist_ok=True)
        p.write_text(body, encoding="utf-8")
        LiveHandler.last_sent = body
        self._broadcast(body)
        self.send_response(204)
        self.send_header("Access-Control-Allow-Origin", "*")
        self.end_headers()

    def do_GET(self) -> None:
        u = urlparse(self.path)
        if u.path == "/mermaid-live":
            self.send_response(200)
            self.send_header("Content-Type", "text/event-stream; charset=utf-8")
            self.send_header("Cache-Control", "no-cache, no-transform")
            self.send_header("Connection", "keep-alive")
            self.send_header("Access-Control-Allow-Origin", "*")
            self.end_headers()
            wfile = self.wfile
            with LiveHandler.clients_lock:
                LiveHandler.clients.add(wfile)
            try:
                text = self._read_file()
                payload = json.dumps({"text": text})
                wfile.write(f"data: {payload}\n\n".encode())
                wfile.flush()
                # manter conexão aberta (heartbeats evitam proxies cortarem)
                while True:
                    time.sleep(25)
                    wfile.write(b": ping\n\n")
                    wfile.flush()
            except (BrokenPipeError, ConnectionResetError, OSError):
                pass
            finally:
                with LiveHandler.clients_lock:
                    LiveHandler.clients.discard(wfile)
            return
        return SimpleHTTPRequestHandler.do_GET(self)

    def guess_type(self, path: str) -> str:
        ext = Path(path).suffix.lower()
        return MIME.get(ext, "application/octet-stream")


def poll_thread(watch: Path) -> None:
    handler = LiveHandler
    handler.watch_file = watch
    while True:
        time.sleep(0.35)
        try:
            st = watch.stat()
            m = st.st_mtime
        except OSError:
            continue
        if handler.last_mtime is None:
            handler.last_mtime = m
            continue
        if m != handler.last_mtime:
            handler.last_mtime = m
            text = watch.read_text(encoding="utf-8")
            if text == handler.last_sent:
                continue
            handler.last_sent = text
            payload = json.dumps({"text": text})
            chunk = f"data: {payload}\n\n".encode()
            with handler.clients_lock:
                dead = []
                for wfile in handler.clients:
                    try:
                        wfile.write(chunk)
                        wfile.flush()
                    except OSError:
                        dead.append(wfile)
                for w in dead:
                    handler.clients.discard(w)


def main() -> None:
    ap = argparse.ArgumentParser()
    ap.add_argument("--file", default="diagram.mmd", type=Path)
    ap.add_argument("--static", dest="static_root", default=".", type=Path)
    ap.add_argument("--port", type=int, default=9876)
    ap.add_argument("--bind", default="127.0.0.1")
    args = ap.parse_args()

    root = args.static_root.resolve()
    watch = args.file.resolve()
    watch.parent.mkdir(parents=True, exist_ok=True)
    if not watch.exists():
        watch.write_text(
            "%%{init: {'theme': 'dark'}}%%\nflowchart TD\n  A[Live] --> B[OK]\n",
            encoding="utf-8",
        )

    LiveHandler.watch_file = watch
    LiveHandler.last_sent = watch.read_text(encoding="utf-8")
    try:
        LiveHandler.last_mtime = watch.stat().st_mtime
    except OSError:
        LiveHandler.last_mtime = None
    os.chdir(root)

    t = threading.Thread(target=poll_thread, args=(watch,), daemon=True)
    t.start()

    httpd = ThreadingHTTPServer((args.bind, args.port), LiveHandler)
    print(
        f"mermaid-live http://{args.bind}:{args.port}/  file={watch}  static={root}",
        flush=True,
    )
    print("  SSE GET /mermaid-live   POST /mermaid-push   abrir página com ?live=1", flush=True)
    httpd.serve_forever()


if __name__ == "__main__":
    main()
